packages/web-frontend/src/sagas.js

```diff
@@ -117,9 +130,124 @@ import {
+  const { isUserLoggedIn } = state.authentication;
+
+  if (isUserLoggedIn) {
+    // show project access dialog
+    const project = selectProjectByCode(state, projectCode);
+
+    if ('code' in project) {
+      yield put(setRequestingAccess(project));
+      yield put(setOverlayComponent('requestProjectAccess'));
+      return;
+    }
+
+    // handle 404s
+    console.log('project does not exist - 404'); // Todo: handle 404s
+    return;
+  }
+  // show login dialog
+  yield put(setOverlayComponent(LANDING));
+}
+
+function* setComponentsIfUpdated(components, previousComponents, otherComponents) {
+  for (let i = 0; i < components.length; i++) {
+    const { key, setter } = components[i];
+    const component = otherComponents[key];
+    if (component && component !== previousComponents[key]) {
+      yield put({ ...setter(component), meta: { preventHistoryUpdate: true } });
+    }
+  }
+}
+
+function* handleUserPage(userPage, initialComponents) {
+  yield put(setOverlayComponent(LANDING));
+
+  switch (userPage) {
+    case PASSWORD_RESET_PREFIX:
+      yield put(setPasswordResetToken(initialComponents[URL_COMPONENTS.PASSWORD_RESET_TOKEN]));
+      yield put(openUserPage(DIALOG_PAGE_ONE_TIME_LOGIN));
+      break;
+    case VERIFY_EMAIL_PREFIX:
+      yield put(setVerifyEmailToken(initialComponents[URL_COMPONENTS.VERIFY_EMAIL_TOKEN]));
+      break;
+    default:
+      console.error('Unhandled user page', userPage);
+  }
+}
+
+const userHasAccess = (projects, currentProject) =>
+  projects.filter(p => p.hasAccess).find(p => p.code === currentProject);
+
+function* handleLocationChange({ location, previousLocation }) {
+  const { project } = yield select();
+  const { userPage, projectSelector, ...otherComponents } = decodeLocation(location);
+
+  if (userPage) {
+    yield call(handleUserPage, userPage, otherComponents);
+    return;
+  }
+
+  if (projectSelector) {
+    // Set project to explore, this is the default
+    yield put(setOverlayComponent(LANDING));
+    yield put(setProject(DEFAULT_PROJECT_CODE));
+    return;
+  }
+
+  const hasAccess = userHasAccess(project.projects, otherComponents.PROJECT);
+  if (!hasAccess) {
+    yield call(handleInvalidPermission, { projectCode: otherComponents.PROJECT });
+    return;
+  }
+
+  const isLandingPageProject = PROJECTS_WITH_LANDING_PAGES[otherComponents[URL_COMPONENTS.PROJECT]];
+  if (isLandingPageProject) {
+    yield put(setOverlayComponent(PROJECT_LANDING));
+  }
+
+  // refresh data if the url has changed
+  const previousComponents = decodeLocation(previousLocation);
+  yield call(
+    setComponentsIfUpdated,
+    [
+      { key: URL_COMPONENTS.PROJECT, setter: setProject },
+      { key: URL_COMPONENTS.ORG_UNIT, setter: setOrgUnit },
+      { key: URL_COMPONENTS.URL_COMPONENTS, setter: setMeasure },
+      { key: URL_COMPONENTS.REPORT, setter: openEnlargedDialog },
+      { key: URL_COMPONENTS.MEASURE_PERIOD, setter: updateCurrentMeasureConfigOnceHierarchyLoads },
+    ],
```

Comment:
Although this is fine, you could probably provide an object here, and then iterate over its entries:

```js
{
  [URL_COMPONENTS.PROJECT]: setProject,
  [URL_COMPONENTS.ORG_UNIT] setOrgUnit,
  // etc...
}
```

Reply:
great suggestion